### PR TITLE
Adding gradient checkpointing to GPT2

### PR DESCRIPTION
**Pull request by TevenLeScao**
This PR adds gradient checkpointing capabilities to GPT-2, imitating the Longformer and Bert checkpointing code. It also disables `find_unused_parameters` in Trainer if the model is using gradient checkpointing, as per #4659 they are incompatible.

### CODE REVIEW

File: src/transformers/configuration_gpt2.py [hunk 1]
```diff
@@ -103,6 +103,8 @@ class GPT2Config(PretrainedConfig):
             :class:`~transformers.GPT2DoubleHeadsModel` and :class:`~transformers.TFGPT2DoubleHeadsModel`.
 
             The dropout ratio to be used after the projection and activation.
+        gradient_checkpointing (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            If True, use gradient checkpointing to save memory at the expense of slower backward pass.
```
Comment (patrickvonplaten) [inline src/transformers/configuration_gpt2.py:107]:
```suggestion
            If :obj:`True`, use gradient checkpointing to save memory at the expense of slower backward pass.
```

File: src/transformers/modeling_gpt2.py [hunk 2]
```diff
@@ -624,16 +623,35 @@ def forward(
...
-                hidden_states,
-                layer_past=layer_past,
-                attention_mask=attention_mask,
-                head_mask=head_mask[i],
-                encoder_hidden_states=encoder_hidden_states,
-                encoder_attention_mask=encoder_attention_mask,
-                use_cache=use_cache,
-                output_attentions=output_attentions,
-            )
+            if getattr(self.config, "gradient_checkpointing", False):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_gpt2.py:626]: I think `if self.config.gradient_checkpointing:` is nicer

File: tests/test_modeling_gpt2.py [hunk 3]
```diff
@@ -355,6 +365,10 @@ def test_gpt2_double_lm_head_model(self):
         config_and_inputs = self.model_tester.prepare_config_and_inputs()
         self.model_tester.create_and_check_double_lm_head_model(*config_and_inputs)
 
+    def test_gpt2_gradient_checkpointing(self):
```
Comment (patrickvonplaten) [inline tests/test_modeling_gpt2.py:368]: Awesome that you add a test!

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me. Left mostly nits.

Would be great if you could run `RUN_SLOW=1 pytest tests/test_modeling_gpt2.py` once to be sure.

File: src/transformers/modeling_gpt2.py (hunk 2 above)
Comment (TevenLeScao) [reply src/transformers/modeling_gpt2.py:626]: Most model `config`s don't actually have this attribute, only the ones that support checkpointing (AFAIK, Bert and Longformer for now) so it's less risky to do things this way.

**Comment by TevenLeScao** [comment]
The slow tests are passing - I've also added a test for generation with checkpointing, although of course to be sure, one should also check the contents of the backwards pass.

File: src/transformers/configuration_gpt2.py [hunk 4]
```diff
@@ -103,6 +103,8 @@ class GPT2Config(PretrainedConfig):
             :class:`~transformers.GPT2DoubleHeadsModel` and :class:`~transformers.TFGPT2DoubleHeadsModel`.
 
             The dropout ratio to be used after the projection and activation.
+        gradient_checkpointing (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            If :obj:`True`, use gradient checkpointing to save memory at the expense of slower backward pass.
```
Comment (sgugger) [inline src/transformers/configuration_gpt2.py:107]: nit: 
```suggestion
            Whether or not to use gradient checkpointing to save memory at the expense of slower backward pass.
```

**Review by sgugger — APPROVED** [review]
Very nice! Thanks a lot for adding this.

File: src/transformers/modeling_gpt2.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_gpt2.py:626]: This is in `modeling_gpt2.py` which only works with `configuration_gpt2.py`. So if you add `gradient_checkpointing` to the config with default = `False` I don't see why this would be risky

**Review by LysandreJik — APPROVED** [review]
LGTM! Thanks for working on it @TevenLeScao!
